Add new sections to SECURITY.md and CONTRIBUTING.md (#1408)
certusone/wormhole: SECURITY.md

    - **Scopes**: *Ethereum Contracts, Solana Contracts, Terra Contracts, Guardian, and Solitaire*
- **[July 1, 2022 - Kudelski](https://storage.googleapis.com/wormhole-audits/2022-07-01_kudelski.pdf)**
    - **Scopes**: *Ethereum Contracts, Solana Contracts, Terra Contracts, and Guardian*

## Rigorous Testing and Review

We believe that rigorous testing that is also transparent is critically important to ensuring the integrity of Wormhole components. It is also critically important that everyone can independantly verify and extend Wormhole test cases, which allow us to further ensure that Wormhole components will operate as expected in both positive and especially negative test case scenarios.

Places to find out more about existing test coverage and how to run those tests:

- **Guardian Node**
    - Tests: `./node/**/*_test.go`

>     - Run: `cd node && go test -v -ldflags '-extldflags "-Wl,--allow-multiple-definition" ' ./...`

Reviewer: petition to get
```
cd node && make test
```
to work

Reviewer 2: Strong concur

- **Ethereum Smart Contracts**
    - Tests: `./ethereum/test/*.[js|sol]`
    - Run: `cd ethereum && make test`
- **Solana Smart Contracts**
    - Tests: `./solana/bridge/program/tests/*.rs`
    - Run: `cd solana && make test`
- **Terra Smart Contracts**
    - Tests: `./terra/test/*`
    - Run: `cd terra && make test`

> We additionally subscribe to a number of linting frameworks, including gosec, golint, cargo check and others to avoid obvious pitfalls and language specific bugs to provide instant feedback to developers on Wormhole.

Author: Removed.

> We additionally subscribe to a number of linting frameworks, including gosec, golint, cargo check and others to avoid obvious pitfalls and language specific bugs to provide instant feedback to developers on Wormhole.

Reviewer: Also, let's run clippy in CI.

    - **Scopes**: *Ethereum Contracts, Solana Contracts, Terra Contracts, Guardian, and Solitaire*
- **[July 1, 2022 - Kudelski](https://storage.googleapis.com/wormhole-audits/2022-07-01_kudelski.pdf)**
    - **Scopes**: *Ethereum Contracts, Solana Contracts, Terra Contracts, and Guardian*

## White-Hat Hacking on Wormhole

We want to lower the bar for White-hat hackers to find security bugs in Wormhole.  Why? The easier we make this process, the more likely it will be for white-hats to find bugs in Wormhole and responsibly disclose them, helping to secure the network.

> Here's a list of strategies that we've found helpful to hackers getting started on Wormhole:

Reviewer 2: ```suggestion
Here's a list of strategies we've found helpful for getting started on Wormhole:
```

- Review the existing unit and integration testing (found in [CONTRIBUTING.md](https://github.com/certusone/wormhole/blob/dev.v2/CONTRIBUTING.md)) and see what we're already testing for.
    * Check out places were we might be missing test coverage entirely.  This could be a ripe spot to look for something we missed.
    * Check out places were we have unit/integration tests, but we lack sufficient [negative test](https://en.wikipedia.org/wiki/Negative_testing) coverage.

> - Review our different smart contract implementations (eg. Solana, EVM, CosmWasm) and attempt to understand how and why they are different.

Reviewer 2: ```suggestion
- Review our different smart contract implementations (eg. Solana, EVM, CosmWasm, Move) and attempt to understand how and why they are different.
```

Just a suggestion to add Move since there is already an existing Aptos PR.

    * Does one chain have a safety check that another chain doesn't?

>     * Does one chain have a specific set of nuances gotchas that that we missed?

Reviewer 2: ```suggestion
    * Does one chain have a specific set of nuances / gotchas that that were missed on another chain?
```

- Consider going beyond the source code

>     * Review the deployed contracts on chain, is some sort of odd that we missed? 

Reviewer 2: ```suggestion
    * Review the deployed contracts on chain. Is something odd that we missed? 
```

We'll continue to iterate on this list of white-hat bootstrap strategies as we grow our lessons learned internally hacking on Wormhole and from other white-hats who have been successful via our bug bounty program.

> It's important to remember that this is an iterative process, if you spend the time to come up with a new test case, but didn't actually find a bug, we'd be extremely appreciative if you'd be willing to send a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) with additional positive and negative test cases.  This process has shown repeatedly to improve your ability to understand Wormhole, and will increase your odds of success.

Reviewer 2: ```suggestion
It's important to remember this is an iterative process. If you spend the time to come up with a new test case, but didn't actually find a bug, we'd be extremely appreciative if you'd be willing to send a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) with additional positive and negative test cases.  This process has shown repeatedly to improve your ability to understand Wormhole, and will increase your odds of success.
```

## Guidance to Chain Integrators

> As the list of chains connected to Wormhole increases, so does the risk that a given connected could introduce risks to the Wormhole network.  As a result, Wormhole does have a built-in safety feature (see [Governor white-paper](https://github.com/certusone/wormhole/blob/dev.v2/whitepapers/0007_governor.md)) to try and reduce the impact of such case.  That said, a defense in depth strategy is required to do as much as possible to secure the network.  As part of this methodology, the Wormhole project recommends that all connected chains current and future implement robust security programs of their own to do their part in managing chain compromise risk to the wormhole network.

Reviewer 2: ```suggestion
As the list of chains connected to Wormhole increases, so does the risk that a given connected could introduce risks to the Wormhole network.  As a result, Wormhole does have built-in safety features (e.g.: [Governor white-paper](https://github.com/certusone/wormhole/blob/dev.v2/whitepapers/0007_governor.md)) to reduce the "blast radius" of such case. That said, a defense in depth strategy is required to do as much as possible to secure the network. As part of this methodology, the Wormhole project recommends that all connected chains current and future implement robust security programs of their own to do their part in managing chain compromise risk to the wormhole network.
```

> Here are a few ways in which connected chains can help can ensure safety of the Wormhole network by maintaining high security standards:

Reviewer 2: ```suggestion
Here are a few ways in which connected chains can maintain high security standards:
```

> - Ensure that all relevant source code is open source.
> - Ensure that all relevant source code is audited by an independant third party and that audit reports are made available to the public.
> - Ensure that all relevant source code is included in a public bug bounty program and that the bounty rewards are sufficiently large to incentivize white-hat mindshare in finding security bugs in your code and responsibly disclosing them.
> - Ensure that all relevant source code makes use of branch protection and have a minimum of one independant reviewer to merge code.
> - Ensure that all relevant source code maintains a [SECURITY.md](https://github.com/certusone/wormhole/blob/dev.v2/SECURITY.md) in the root of the repository (like this one) to offer guidance and transparency on security relevant topics.
> - Ensure that all relevant source code has sufficient unit test and integration test coverage (including negative tests), which is run on every commit via continuous integration.  Additionally, ensure that the results of those test runs are visible to the public.
> - Ensure that the Wormhole team has sufficient contact information and an associated call or page tree to reach you in the event of a security incident.
> - Ensure that Wormhole has the full upgrade authority on relevant bridge contracts to act quickly in the case of security incident.
> - Ensure that you have an established incident response program in place, with established patterns and playbooks to ensure deterministic outcomes for containment.
> - Ensure that when security issues do occur, that the chain makes every attempt to inform affected parties and lead with transparency.

Reviewer 2: I tried to cut out the repetition `Ensure that all relavant source code...` and clean this up a tad. Let me know if this is good or not.

```suggestion
For source code ensure relevant bits are:
- All open source
- Audited by an independent third party with public audit reports
- Included in a public bug bounty program. The bounty rewards should be sufficiently large to incentivize white-hat mindshare in finding security bugs and responsibly disclosing them
- Version control systems contain adequate access controls and mandatory code review (e.g.: In github, use of branch protection and a minimum of one independent reviewer to merge code)
- Maintaining a [SECURITY.md](https://github.com/certusone/wormhole/blob/dev.v2/SECURITY.md) in the root of the repository (like this one) to offer guidance and transparency on security relevant topics
- Includes sufficient unit and integration test coverage (including negative tests), which are run on every commit via continuous integration. Ensure that the results of those test runs are visible to the public

Additionally, ensure:

- The Wormhole team has sufficient contact information and an associated call or page tree to reach you in the event of a security incident.
- That Wormhole has the full upgrade authority on relevant bridge contracts to act quickly in the case of a security incident.
- You have an established incident response program in place, with established patterns and playbooks to ensure deterministic outcomes for containment.
- When security issues do occur, that the chain makes every attempt to inform affected parties and leads with transparency.
```
